### ISSUE TITLE
Tryout refactoring model class

### ISSUE THREAD

**Issue by stijnvanhoey** [UGentBiomath/COVID19-Model#42]
The current `SEIRSAgeModel` class has a lot of overhead when extending it, with a lot of initializations on differen locations in the class,... Moreover, a `run_epoch` wrapped by a `run` method which is wrapped by a `sim` method in the same class. We would rather split up the class, had some discussion and did some dummy testing (see also [this gist](https://gist.github.com/stijnvanhoey/0b6a89c1ab12e4c9a4bff123cdf281b4) for a conceptual test).

- have a `BaseModel` class which handles most of the work in the background (cfr. all the numpy arrays that are initialized, e.g. [here](https://github.com/UGentBiomath/COVID19-Model/blob/master/src/covid19model/models/models.py#L456-L498)) in order to have a new model definition defined as a subclass with a single method `integrate`, e.g. `SIR` model:

```python
# User (model developer) specifies...

# ... state transitions in a subclass
class SIR(BaseModel):

    # state variables and parameters
    state_names = ['S', 'I', 'R']
    parameter_names = ['beta', 'gamma']
    
    @staticmethod
    def integrate(t, S, I, R, beta, gamma):  # All variables and parameters as it is now... will be long list or arguments(!)
        """Basic SIR model"""
        N = S + I + R
        dS = -beta*S*I/N
        dI = beta*S*I/N - gamma*I
        dR = gamma*I
        
        return dS, dI, dR

# ... parameters and initial conditions
parameters = {"beta": 0.5, "gamma": 0.3}
initial_states = {"S": 7900000, "I": 10, "R": 0}

# -> user initiates the model
sir_model = SIR(initial_states, parameters)

# -> user runs a simulation for a defined time period
time = [0, 150]
t, output = sir_model.sim(time)

# users sees dictionaries to get variables, e.g. for plotting
fig, ax = plt.subplots()
ax.plot(t, output["S"])
ax.plot(t, output["I"])
ax.plot(t, output["R"])

# -> user can do fit, mc,... using the model class instance `sir_model` 
```

As such, we can split up model development (the equations -> subclasses) and model 'handling' (all the array-bookkeepping,... -> base class).

@jorisvandenbossche can you have a check on the [gist](https://gist.github.com/stijnvanhoey/0b6a89c1ab12e4c9a4bff123cdf281b4) and see what you think of the exposed user API? We then need to extend this dummy test to be able to handle the stratification (with interaction matrix) of the variables (e.g. by age classes). 

Note, we do not need to define the Monte Carlo (which is running the same model instance for multiple parameters) and `epoch` concept inside the function, as both can be defined by wrapping the class instance (more general applicable for other models as well). For the `epoch` concept we can either concatenate multiple model runs of the same instance (which is more or less the same as current implementation) or using the `event` concept (need to be tested though).

**Comment by JennaVergeynst** [comment]
@twallema, what do you think of this refactoring? Is it clear to you or should we have a short meeting with Stijn/Joris to make everything clear? I think this could facilitate working with the model a lot!

**Comment by twallema** [comment]
This seems like a good idea! I took a look at the gist but some things still seem a bit abstract to me, so a meeting may be necessary.

**Comment by stijnvanhoey** [comment]
After meeting, we agree on the format and continue the work as such. short report:

### Meeting

NEXT STEP 1
- stratificatie op SIR model uittesten
- SEIRage model in nieuwe structuur
- genereren van de xarray structuur -> plot functie op basis daarvan

NEXT STEP 2
- MC/fit koppeling/wrapping doen

### Model

* `matmul` element moet er nog altijd inzitten, moet de modelontwikkelaar doen -> te documenteren
* outputs te definieren in xarray: 
    * dimensies: stratificatie (bv. leeftijd), tijd en herhalingen (MC), variabele/state  
    * attributen te definiëren voor bv. de parameters
    * voor de basis-gebruiker xarray af te schermen
    => plot functies hiervoor te herschrijven
    => alles samen te brengen als een model-object zodat je alles van model reproduceren
* ivm vertaalslag: variabelen van model versus vergelijken met de metingen 
    => niet op zich als afzonderlijk veel te definiëren, is tyisch niet meer dan som van enkele variabelen
    => voorlopig vooral belangrijk dat ze als namen worden gedefinieerd; dat verduidelijkt al veel
* optimalisaties:
    => op MC/PSO level snel winst door daar parallel te werken
    => optioneel: mumba per simulatie
    
### DATA

to be listed in `src/covid19model/data`:

- mobility data Google
- sciensano data

**Comment by stijnvanhoey** [comment]
Minimal working version of the same concept with stratification (e.g. age,...) https://gist.github.com/stijnvanhoey/0b6a89c1ab12e4c9a4bff123cdf281b4#file-refactor_stratified-ipynb

**Comment by stijnvanhoey** [comment]
@jorisvandenbossche, I reimplemented the plotting, see https://github.com/stijnvanhoey/COVID19-Model/blob/refactor/src/covid19model/visualization/output.py and as I'm now also looking at the optimization implementation, I do think the extension of out class-definition with an 'aggregated variables' method or description would be beneficial for these purposes.  

@twallema would be good to have your input here as well, so let me rephrase how the refactoring changes the way a model is setup by a _model developer_ (i.e. you ;-)).
The layout of the class to create a new model is currently (see also the [SEIRSAge model](https://github.com/stijnvanhoey/COVID19-Model/blob/refactor/src/covid19model/models/models.py#L57-L114) which replicates the previous implementation `SEIRSAgeModel`): 

- `state_names` with the state variables of the model
- `parameter_names` with the regular parameters of the model
- `parameters_stratified_names` with the stratified parameters of the model
- `stratification` with the name of the stratification array
- an `integrate` staticmethod with the model equations itself

As a _model developer_ creating a new model  would come down to writing just these elements, which simplifies the management of different models a lot compared to the current approach:

```
class SEIRSAge(BaseModel):
    """Biomath SEIRS model"""

    # ...state variables and parameters
    state_names = ['S', ..., 'RQ']
    parameter_names = ['beta', ..., 'dq']
    parameters_stratified_names = ['h',..., 'mi']
    stratification = 'Nc'

    # ..transitions/equations
    @staticmethod
    def integrate(t, S, ...,  beta, ..., Nc):
        """My SEIR implementation"""

        # Model equations
        dS = ...
        return (dS, ..., dRQ)
```

However, for plotting purposes and optimization, combinations of these variables are used and instead of 'hardcoding' them in each of these functions (see [this example](https://github.com/stijnvanhoey/COVID19-Model/blob/refactor/src/covid19model/visualization/output.py#L63-L65)), maybe we could extend the class with an additional method or attribute to make these variable combinations (e.g. `C_total`, `I_total`, `H`) easier to access for the user and accessible for each of the methods. It would also replace the [`positions`](https://github.com/stijnvanhoey/COVID19-Model/blob/refactor/src/covid19model/models/models.py#L750) concept of the current implementation. 

Different options do exist on what kind of post-calculations are possible (e.g.hard-coded predefined combinations, a matrix to multiply with or just 'any custom-function'), but I would propose a pragmatic approach in line with the current way these aggregations are used for plotting and parameter optimization - consider [`H` as an example](https://github.com/stijnvanhoey/COVID19-Model/blob/refactor/src/covid19model/models/models.py#L629) - :

- an aggregation is always a sum of 2 or more variables (sum of counts)
- sum can be made element-wise for each time step of the model simulation output

However, I see different options, with respect to the implementation. We can either opt to: 

1. always save all variables of the model + the additional once from the aggregations (e.g. the SEIRSAge model 18 states + H, C_toal and I_total.
2.use the observed states  as a way to define and/or limit the output of the model to the use. 

In other implementations (e.g. optimizations, data assimilation), a typical pattern is to use a 'selection' matrix, which is by default the identity matrix corresponding to 'selecting' all variables as such. By altering this array, other combinations can be created. Let me make this clear with the following example: We have 3 variables (rows: S, I, R) and 4 timesteps (columns):

```
array([[ 1,  2,  3,  4],
          [ 5,  6,  7,  8],
          [ 9, 10, 11, 12]])
```

Multiplication of identity matrix with the model output array, returns the array as such, we are interested in S, I, R as such:
```
>>> np.identity(3) @ output
array([[ 1,  2,  3,  4],
          [ 5,  6,  7,  8],
          [ 9, 10, 11, 12]])
```

But instead of the identify matrix, we could use another matrix:

```
>>> select_matrix = np.array([[1., 0.,  0.],    # interested in S
                                                [0., 1., 1.],     # interested in sum of I and R
                                                [1., 1., 1.]])   # interested in sum of S, I and R
>>> select_matrix @ output
>>>  array([[ 1.,  2.,  3.,  4.],
                   [14., 16., 18., 20.],
                   [15., 18., 21., 24.]])
```

In the same way, we could just select a single variable:
```
>>> select_matrix = np.array([[1., 0.,  0.]]) # interested in S
>>> select_matrix @ output
>>>  array([[1., 2., 3., 4.]])

>>> select_matrix = np.array([[0., 1.,  1.]]) # interested in 
>>> select_matrix @ output
>>>  array([[1., 2., 3., 4.]])
```

This enables us to create a lot of flexibility and at the same time limit the number of states to store in memory (e.g. we could limit it down to just a few states and a few aggregations.  
I do think this pattern is very useful to implement, but the main question is __how/where we integrate this into the class__? I see different options:

- We could do it for each functionality separately, e.g. an input of the fit-method (to create/select the states to fit with), an input to the MC (to only store the relevant variables to limit memory usage) and each plot function as such. And do not make it part of the model itself. 
- We could use it as part of creating a new model instance,  which is by default the identity matrix (just select all variables) but can be provided by the user to switch this setting: 
  
  ```
  observation_array = np.array([[1., 0.,  0.]]) # interested in S only
  sir_model = SIR(initial_states, parameters, observation_array)
  ```
- We could use it as part of creating a new model instance, but ignore the array approach and make it more 'descriptive', e.g.

  ```
  observation_array = {'S' : 'S',                       # interested in S
                                    'total_N': ['S', 'I', 'R']}   # interested in S+I+R
  sir_model = SIR(initial_states, parameters, observation_array)
  ```

Hope this is clear and input is welcome. 

__Note:__  the 'observed_matrix' is not the same as the `weights` matrix as part of the [current fit implementation ](https://github.com/stijnvanhoey/COVID19-Model/blob/refactor/src/covid19model/models/models.py#L754) , which is applied to the model residual (and putting this to zero is shortcut to ignore certain measurements, whereas it would be more clear just not incorporating the variable in such a situation).

**Comment by jorisvandenbossche** [comment]
@stijnvanhoey thanks for writing that down, will tinker on it a bit. I think the main question is: how flexible do those combinations *need* to be in practice? If it is only sums, then something like your last code snippet would easily work. But that might also be too limiting?

@twallema note that there is already an open PR with the refactored model. At this point I think it would be useful that you can test drive this a bit.

**Comment by jorisvandenbossche** [comment]
@stijnvanhoey related to the idea of "observed variables" that can automatically included in the resulting xarray.Dataset output. In #113, we ran into a case that is not representable as the algebraic combination of one of the existing states (https://github.com/UGentBiomath/COVID19-Model/pull/113/files#diff-7a10a383fb174c3a44fbe9452b9eca5cR166-R167). 
Basically we want to keep track of the "new hospitalizations" per day (the flow from "Infected" to "Hospital" in https://ugentbiomath.github.io/COVID19-Model/models.html). But since each of those states have multiples ins/outs, you can't calculate this specific flow "after the facts".

**Comment by JennaVergeynst** [comment]
I think the refactoring is completed and in full use... @twallema are we at the point to clean up redundant classes and functions in the models.py file? (as to decrease its size from 1600 to 300 lines? :-D )

**Comment by twallema** [comment]
@JennaVergeynst , I'll take care of this today. The controller code must be ported.

**Comment by twallema** [comment]
@JennaVergeynst I removed everything non-refactored from `models.py`, MPC code was copied to `src/covid19model/model/MPC.py`
